### DISCUSSION TITLE
Why does order of children completing affect whether or not parent onDone is triggered?

### DISCUSSION THREAD

**Discussion by Raithybabes** [statelyai/xstate#3639] [Q&A]
### Setup
For a parallel state which has onDone
and which has child states which are a mix of compound and parallel
and all sequences contain a final state.

### Happy outcome
If the 'parallel' type children are done before the last 'compound' type child reaches final, then the parent onDone is triggered.

### Unhappy outcome
But if a 'parallel' type child is the last to become done/final, then the parent onDone is not triggered.

### Workaround
One workaround is to encapsulate 'parallel' type children in a compound container and transition onDone to a 'final' type sibling (convert 'done parallel' to 'final compound').

### Reproduction
Finish1 followed by Finish2 = success. 
Finish2 followed by Finish1 = fail.

```
import { createMachine, MachineConfig } from "xstate";

const cfg: MachineConfig<any, any, any> = {
  id: "Order of finishing matters",
  initial: "started",
  states: {
    started: {
      onDone: "end",
      type: "parallel",
      states: {
        parallel_child: {
          type: "parallel",
          states: {
            must_finish_first: {
              initial: "pending",
              states: {
                pending: {
                  on: {
                    finish1: "final",
                  },
                },
                final: {
                  type: "final",
                },
              },
            },
          },
        },
        compound_child: {
          initial: "pending",
          states: {
            pending: {
              on: {
                finish2: "final",
              },
            },
            final: {
              type: "final",
            },
          },
        },
      },
    },
    end: {},
  },
};

createMachine(cfg);
```

Is this by design, or a bug?
Is it known?
Has fixing it already been discounted?
_(I have checked open and closed issues & discussions but couldn't find a match for this)_

Thanks for any clarification you can offer :)

**Comment by ChrisShank** [comment]
This is a good question! XState (for the most part) follows the W3C SCXML specification. While I don't remember off the top of my head, XState implements the semantics for `done.state.<id>` events outlined [here](https://www.w3.org/TR/scxml/#enterStates). The main culprit is determining whether a parallel state is in a final state is not recursive it only checks one up one level of hierarchy. Thus why `finish1` is not triggering the final state to complete. 

```
if isFinalState(s):
    if isSCXMLElement(s.parent):
        running = false
    else:
        parent = s.parent
        grandparent = parent.parent
        internalQueue.enqueue(new Event("done.state." + parent.id, s.donedata))
        if isParallelState(grandparent):
            if getChildStates(grandparent).every(isInFinalState): // 
                internalQueue.enqueue(new Event("done.state." + grandparent.id))
```

```
function isInFinalState(s):
    if isCompoundState(s):
        return getChildStates(s).some(lambda s: isFinalState(s) and configuration.isMember(s))
    elif isParallelState(s):
        return getChildStates(s).every(isInFinalState)
    else:
        return false
```

For the most part (although I can think of some exceptions), I think it is not necessary for a parallel state to have a region that is also parallel. For the examples below are equivalent:

```ts
{
  initial: 'A',
  states: {
    A: {
      type: 'parallel'.
      states: {
        B: {
          type: 'parallel'.
          states: {
            C: {},
            D: {}.
        },
        E: {}
      }
    }
  }
}
```

```ts
{
  initial: 'A',
  states: {
    A: {
      type: 'parallel'.
      states: {
        B: {},
        C: {},
        D: {}.
        E: {}
      }
    }
  }
}
```
(P.S. Although I cant remember where this has been discussed before! Probably on the xstate discord.)

**Reply by ChrisShank** [reply]
Here is one related conversation on discord: https://discord.com/channels/795785288994652170/964555380489338950/964555381990899732

**Reply by Raithybabes** [reply]
Thanks @ChrisShank
I haven't lifted the lid and reasoned about SCXML to explain why this works the way it does - I should one day!
I was learning by experimenting, and this has clearly led to some misunderstanding and dashed expectations.

If I'm reading that enterStates snippet correctly then entering a final state can trigger parent and grandparent as being 'done', whereas I was expecting it to continue bubbling.
Thinking aloud here... a further layer of compound nesting would stop my parallel great-grandparent of becoming done??
I can sort of see how this has come about, because "semantics".

But it still doesn't feel right. My parallel child needs to be 'done' so that its parallel parent can become 'done'. It's not that the child done-ness doesn't matter. It comes across to the layman as flaky that the parent done-ness is dependent on the _order_ of arriving in the correct set of child states.

Thanks also for the discord link (now joined) in which davidkpiano says he'd "love to deviate from SCXML" on this matter. I've lauded xstate for being a reference implementation of the state machine standard, but that conformity comes with a cost. I recall seeing a discussion in an issue (no link, sorry) where a question was raised as to whether the SCXML specification is really complete!

I shall keep my misinformed opinions to myself for now 😆 but I will certainly try to make time to become involved in xstate beyond being a consumer.

**Reply by davidkpiano** [reply]
It is strange behavior, and I'd be happy to revisit this, because the mental model should be that the parallel state is done if all regions are done; as simple as that.

Just checked this in v5, which more faithfully implements the SCXML algorithm, and it's the same story.

**Reply by Raithybabes** [reply]
Great stuff. As a layman consumer, I agree that the mental model would be improved without such arcane knowledge. But maybe I just haven't even achieved layman-consumer level yet 😆 

What is required to revisit this @davidkpiano ?
Does the spec need to be reviewed to see if the current interpretation is off?
Or is the current understanding of the spec accepted, but a decision needs making to intentionally deviate from it? By whom?

I fear I've too much ground to cover before I could usefully contribute but I'd like to help if possible 🤷

**Reply by Raithybabes** [reply]
In [final](https://www.w3.org/TR/scxml/#final) the spec says:
_"When the state machine enters the final child of a state element, the SCXML Processor must generate the event done.state.id after completion of the onentry elements, where id is the id of the parent state. Immediately thereafter, **if the parent state is a child of a parallel element, and all of the parallel's other children are also in final states, the Processor must generate the event done.state.id where id is the id of the parallel element**."_

The interpreter seems to me to literally apply this. If the grandparent of a final state is parallel then it does make that state final too, and trigger onDone. Whereas if the final parent is a compound state within another compound state within a parallel state (don't ask me why but [I believe it is allowed](https://www.w3.org/TR/scxml/#N10192)) then the parallel is not done. The interpreter only raises done for parent and grandparent.

However, for [Parallel States](https://www.w3.org/TR/scxml/#N10199) the spec says:
_"**When all of the children reach final states**, the parallel element itself is considered to be in a final state, and a completion event done.state.id is generated, where id is the id of the parallel element."_
parallel regions may be directly nested. From the parallel state perspective its 'final' children may be other parallel states.

The final state behaviour is described in an imperative way, and seems to be how the xstate interpreter has been implemented.

Whereas the parallel state behaviour is described in a more declarative way, but appears not to be implemented.

How have the team previously reconciled these different approaches in the specification? Is there a clearer explanation elsewhere which narrows the interpretation?

**Reply by Andarist** [reply]
> From the parallel state perspective its 'final' children may be other parallel states.

I don't think that's necessarily true. I'm not a native speaker but I can see the quoted part of the spec being interpreted both ways.

However, in general - I agree with you. This should be recursive.

**Reply by Raithybabes** [reply]
I was trying to distil some inarguable logic from the specification, which otherwise does seem to leave itself quite open to interpretation in places. The following seems quite clear to me, but maybe it is undermined elsewhere in the spec? When arguing a point one tends to stop looking when one find things that support it 😄 

Parallel element definitely may have parallel elements as children. https://www.w3.org/TR/scxml/#parallel describes that relationship as "Defines a nested set of parallel regions"

I think my previously quoted bit about Parallel States is clear about "all of the children", affecting the "element itself" and the expected outcome of "completion event is generated".

So that does sound, to me, that the specification agrees with our opinion that the done event should bubble up and be raised for ancestor parallel states. We all seem to agree on that, but I got the impression that it was considered to be _against_ the spec. I guess I'm just trying to convince people that it is in agreement with the spec, in case that helps get it done 🤷 

Please can you tell me - is it the "done.state.id" event that becomes `onDone` on a parallel state?

**Reply by ChrisShank** [reply]
> is it the "done.state.id" event that becomes onDone on a parallel state?

xstate does use syntactic sugar for this

**Reply by Raithybabes** [reply]
Thanks. Having convinced myself that `done.state.id` should be raised for a parallel state when its children become final, I just wanted to make sure that does actually equate to `onDone`. Otherwise I've been barking up the wrong tree!

I just wish I was in a position to suggest how to implement this. If I start reading the code now then maybe in a year or two... 🤯
